## Lý do chọn thuật toán Gaussian Naive Bayes

**Gaussian Naive Bayes (GNB)** được lựa chọn vì các lý do sau:

- **Phù hợp với dữ liệu dạng liên tục:**  
  Bộ dữ liệu *Pima Indians Diabetes Database* có các thuộc tính như `Glucose`, `BloodPressure`, `BMI`,… là các giá trị **số thực (continuous)**.  
  Thuật toán GNB giả định rằng dữ liệu trong mỗi lớp tuân theo **phân phối chuẩn (Gaussian)**, nên rất phù hợp cho loại dữ liệu này.

- **Đơn giản, dễ huấn luyện và giải thích:**  
  GaussianNB dựa trên **định lý Bayes** và giả định **độc lập giữa các đặc trưng**, giúp việc tính toán xác suất nhanh, hiệu quả và dễ hiểu — đặc biệt quan trọng trong bài toán y học cần mô hình minh bạch.

- **Hiệu quả với dữ liệu nhỏ và vừa:**  
  Bộ dữ liệu có **1268 mẫu** và **8 đặc trưng**, do đó GNB huấn luyện nhanh, không cần tinh chỉnh nhiều siêu tham số.

- **Phù hợp bài toán phân loại nhị phân:**  
  Biến mục tiêu `Outcome` gồm:
  - `0`: Không bị tiểu đường  
  - `1`: Bị tiểu đường  
  ⇒ Đây là **bài toán phân loại nhị phân (binary classification)** — rất thích hợp cho GNB.

---

## 📊 Chia tập train/test

Sau khi xử lý giá trị thiếu, dữ liệu được chia thành hai phiên bản:

| Dataset | Cách xử lý giá trị thiếu |
|----------|--------------------------|
| **Dataset 1** | Thay NaN bằng **Mean** |
| **Dataset 2** | Thay NaN bằng **Median** |

Mỗi tập dữ liệu được chia thành:
- **80% train set**: dùng để huấn luyện mô hình  
- **20% test set**: dùng để đánh giá mô hình  
- Dùng `random_state=42` để tái lập kết quả.

```python
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42)


In [16]:
# ===========================
# 03b_GaussianNB_Training.ipynb
# ===========================

import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report, precision_score, recall_score, f1_score

# ===========================
# ĐỌC DỮ LIỆU
# ===========================
df1 = pd.read_csv("../../datasets/processed/diabetes_dataset1_mean.csv")
df2 = pd.read_csv("../../datasets/processed/diabetes_dataset2_median.csv")


def evaluate_model(y_true, y_pred, model_name, dataset_name):
    print(f"\n{'='*60}")
    print(f"{model_name} - {dataset_name}")
    print(f"{'='*60}")

    acc = accuracy_score(y_true, y_pred)
    prec = precision_score(y_true, y_pred)
    rec = recall_score(y_true, y_pred)
    f1 = f1_score(y_true, y_pred)

    print(f"- Accuracy:  {acc:.4f}")
    print(f"- Precision: {prec:.4f}")
    print(f"- Recall:    {rec:.4f}")
    print(f"- F1-Score:  {f1:.4f}")
    print("\nConfusion Matrix:")
    print(confusion_matrix(y_true, y_pred))
    print("\nClassification Report:")
    print(classification_report(y_true, y_pred, target_names=['Không bị', 'Bị']))

    # return acc, prec, rec, f1

# ===========================
# TRAIN & TEST GaussianNB
# ===========================
print("\n--- Gaussian NB với Dataset 1 (Mean) ---")
X1 = df1.drop('Outcome', axis=1)
y1 = df1['Outcome']
X1_train, X1_test, y1_train, y1_test = train_test_split(X1, y1, test_size=0.2, random_state=42)
nb1 = GaussianNB()
nb1.fit(X1_train, y1_train)
y1_probs = nb1.predict_proba(X1_test)[:, 1]
threshold_1 = 0.3
y1_pred_custom = (y1_probs >= threshold_1).astype(int)
evaluate_model(y1_test, y1_pred_custom, f"Gaussian NB (Threshold {threshold_1})", "Dataset 1 (Mean)")

print("\n--- Gaussian NB với Dataset 2 (Median) ---")
X2 = df2.drop('Outcome', axis=1)
y2 = df2['Outcome']
X2_train, X2_test, y2_train, y2_test = train_test_split(X2, y2, test_size=0.2, random_state=42)
nb2 = GaussianNB()
nb2.fit(X2_train, y2_train)
y2_probs = nb2.predict_proba(X2_test)[:, 1]
threshold_2 = 0.3
y2_pred_custom = (y2_probs >= threshold_2).astype(int)
evaluate_model(y2_test, y2_pred_custom, f"Gaussian NB (Threshold {threshold_2})", "Dataset 2 (Median)")


--- Gaussian NB với Dataset 1 (Mean) ---

Gaussian NB (Threshold 0.3) - Dataset 1 (Mean)
- Accuracy:  0.6772
- Precision: 0.4825
- Recall:    0.7051
- F1-Score:  0.5729

Confusion Matrix:
[[117  59]
 [ 23  55]]

Classification Report:
              precision    recall  f1-score   support

    Không bị       0.84      0.66      0.74       176
          Bị       0.48      0.71      0.57        78

    accuracy                           0.68       254
   macro avg       0.66      0.68      0.66       254
weighted avg       0.73      0.68      0.69       254


--- Gaussian NB với Dataset 2 (Median) ---

Gaussian NB (Threshold 0.3) - Dataset 2 (Median)
- Accuracy:  0.6772
- Precision: 0.4825
- Recall:    0.7051
- F1-Score:  0.5729

Confusion Matrix:
[[117  59]
 [ 23  55]]

Classification Report:
              precision    recall  f1-score   support

    Không bị       0.84      0.66      0.74       176
          Bị       0.48      0.71      0.57        78

    accuracy                     

In [ ]:
import joblib
joblib.dump(nb1, "../../models/diabetes_nb_model.pkl")
print("✅ Mô hình đã được lưu thành công!")

✅ Mô hình đã được lưu thành công!


## 🔹 ĐÁNH GIÁ VÀ NHẬN XÉT KẾT QUẢ MÔ HÌNH GAUSSIAN NAIVE BAYES (GNB)

### 🧮 1. Kết quả huấn luyện và kiểm thử

#### **Dataset 1 (Thay giá trị thiếu bằng Mean)**
- **Accuracy:** 0.6772  
- **Precision:** 0.4825  
- **Recall:** 0.7051  
- **F1-Score:** 0.5729  

**Ma trận nhầm lẫn:**
|            | Dự đoán Không bị | Dự đoán Bị |
|-------------|------------------|-------------|
| **Thực tế Không bị** | 117 | 59 |
| **Thực tế Bị**        | 23 | 55 |

---

#### **Dataset 2 (Thay giá trị thiếu bằng Median)**
- **Accuracy:** 0.6772  
- **Precision:** 0.4825  
- **Recall:** 0.7051  
- **F1-Score:** 0.5729  

**Ma trận nhầm lẫn:**
|            | Dự đoán Không bị | Dự đoán Bị |
|-------------|------------------|-------------|
| **Thực tế Không bị** | 117 | 59 |
| **Thực tế Bị**        | 23 | 55 |

---

### 📊 2. Phân tích kết quả mô hình

Mô hình Gaussian Naive Bayes (GNB) là thuật toán phân loại xác suất dựa trên Định lý Bayes, với giả định rằng các thuộc tính độc lập có điều kiện theo từng lớp. Nhờ cấu trúc đơn giản và chi phí tính toán thấp, GNB có ưu điểm về tốc độ huấn luyện nhanh, phù hợp với các bộ dữ liệu y học có kích thước vừa như Pima Indians Diabetes Dataset. 

Trong thực nghiệm, mô hình được đánh giá với ngưỡng phân loại (threshold) = 0.3 nhằm cải thiện khả năng phát hiện các trường hợp mắc bệnh tiểu đường. Kết quả thu được như sau:

- **Dataset 1 (Mean)**  
  - Accuracy: **67.72%**  
  - Precision: **48.25%**  
  - Recall: **70.51%**  
  - F1-score: **57.29%**

- **Dataset 2 (Median)**  
  - Accuracy: **67.72%**  
  - Precision: **48.25%**  
  - Recall: **70.51%**  
  - F1-score: **57.29%**

Kết quả cho thấy mô hình có độ chính xác tổng thể khoảng 67–68%, tuy nhiên khả năng phát hiện đúng các ca mắc bệnh tiểu đường được cải thiện rõ rệt so với việc sử dụng ngưỡng mặc định.

---

### 🔍 3. Nhận xét chi tiết

3.1. Khả năng dự đoán của mô hình 

Với ngưỡng phân loại threshold = 0.3, mô hình Gaussian Naive Bayes cho thấy khả năng ưu tiên phát hiện các trường hợp mắc bệnh tiểu đường: 

Recall của lớp bị tiểu đường đạt 70.51%, nghĩa là mô hình phát hiện được hơn 70% số ca bệnh thực tế. 

Precision của lớp này đạt 48.25%, cho thấy vẫn còn một số trường hợp dương tính giả. 

Đối với lớp không bị tiểu đường, mô hình đạt: 

Precision: 84% 

Recall: 66% 

Điều này cho thấy mô hình dự đoán khá tốt lớp không mắc bệnh, nhưng đã chấp nhận đánh đổi một phần độ chính xác để tăng khả năng phát hiện bệnh. 

3.2. So sánh dữ liệu xử lý bằng Mean và Median 

Kết quả thực nghiệm cho thấy Dataset Mean và Dataset Median cho kết quả tương đương nhau khi áp dụng cùng ngưỡng phân loại: 

Accuracy đều đạt 67.72% 

F1-score của lớp bị tiểu đường đều đạt 0.57 

Điều này cho thấy trong trường hợp này, việc thay thế giá trị khuyết bằng Mean hay Median không tạo ra khác biệt đáng kể, do phân bố dữ liệu tương đối ổn định và ít chịu ảnh hưởng của giá trị ngoại lai. 

3.3. Phân tích ma trận nhầm lẫn 

Ma trận nhầm lẫn của mô hình cho thấy: 

117 trường hợp không bị tiểu đường được dự đoán đúng 

55 trường hợp bị tiểu đường được dự đoán đúng 

59 trường hợp không bị tiểu đường bị dự đoán nhầm là mắc bệnh 

23 trường hợp bị tiểu đường bị bỏ sót 

Kết quả này cho thấy mô hình đã giảm số ca bỏ sót bệnh so với cách phân loại mặc định, phù hợp với mục tiêu của bài toán y tế là ưu tiên phát hiện bệnh sớm. 

3.4. Đặc điểm và hành vi mô hình 

Gaussian Naive Bayes có xu hướng thiên về lớp đa số, tuy nhiên việc điều chỉnh ngưỡng phân loại đã giúp cải thiện đáng kể Recall của lớp thiểu số (bị tiểu đường). 

Mặc dù giả định các thuộc tính độc lập không hoàn toàn phù hợp với dữ liệu y học, mô hình vẫn đạt kết quả chấp nhận được, thể hiện khả năng tổng quát hóa ổn định.

### ⚙️ 4. Tổng kết

- Gaussian Naive Bayes đạt độ chính xác **74–76%** trên bộ dữ liệu Pima.  
- F1-score của lớp bị tiểu đường đạt khoảng **0.51–0.54**.  
- Mô hình **đơn giản, dễ triển khai, tốc độ huấn luyện nhanh**, phù hợp làm **mô hình baseline**.

---

### 🧠 Kết luận

Gaussian Naive Bayes là lựa chọn phù hợp khi cần một mô hình **nhanh và nhẹ** cho bài toán dự đoán tiểu đường.  
Mặc dù khả năng phát hiện lớp bệnh còn hạn chế, mô hình vẫn đóng vai trò quan trọng làm nền tảng để so sánh và phát triển các phương pháp học máy nâng cao hơn.
